stingray/pulse/overlapandsave/ols.py

```diff
@@ -0,0 +1,355 @@
+"""
+The code in this file was copied from
+https://github.com/fasiha/array_range
+https://github.com/fasiha/nextprod-py
+https://github.com/fasiha/overlap_save-py
+
+Thanks to Ahmed Fasih for releasing this to the larger public with the
+Unlicense.
+
+# Fast-convolution via overlap-save: a partial drop-in replacement for scipy.signal.fftconvolve
+
+Features:
+
+- 1D and 2D (both tested) and higher (untested) arrays
+  - (Currently unsupported is convolving different-dimensional signals)
+- Real and complex arrays
+- Memory-mapped input/outputs fully supported (and tested)
+- Supports alternative FFT engines such as PyFFTW
+- Supports reflect-mode (signal assumed to reflect infinitely, instead of 0
+  outside its support; useful for avoiding edge effects)
+- Relatively straightforward to paralellize each step of the algorithm
+- Extensively unit-tested
+
+When it can be used as a drop-in replacement for `fftconvolve`:
+
+- when you call `fftconvolve` with `mode='same'` and `axes=None`
+- [See docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.fftconvolve.html)
+
+Example (culled from unit tests):
+```py
+import numpy as np
+from scipy.signal import fftconvolve
+
+# Generate a 100x100 signal array and a 10x10 filter array
+nx = 100
+nh = 10
+x = np.random.randint(-10, 10, size=(nx, nx)) + 1.0
+h = np.random.randint(-10, 10, size=(nh, nh)) + 1.0
+
+# Compute the linear convolution using the FFT, keeping the center 100x100 samples
+expected = fftconvolve(x, h, mode='same')
+
+# Use overlap-save, computing the output in 6x5-sized chunks. Instead of one huge FFT, we do a
+# several tiny ones
+from ols import ols
+actual = ols(x, h, [6, 5])
+
+# The two will match
+assert np.allclose(expected, actual)
+```
+"""
+
+import numpy as np
+
+from itertools import product
+from typing import List, Iterable, Tuple
+
+from typing import List
+import math
+
+
+def flip_all(array):
+    """Flip all array dimensions (compatibility with old numpy)
+
+    Substitutes for np.flip(array, axis=None) introduced in Numpy 1.15
+
+    Examples
+    --------
+    >>> array = np.array([[[0, 1], [2, 3]], [[4, 5], [6, 7]]])
+    >>> flipped_array = np.array([[[7, 6], [5, 4]], [[3, 2], [1, 0]]])
+    >>> np.all(flipped_array == flip_all(array))
+    True
+    """
+    for dim in range(len(array.shape)):
+        array = np.flip(array, dim)
+    return array
+
+
+try:
+    from numpy import flip
+    flip(np.zeros((0, 0)))
+except TypeError:
+    flip = flip_all
+
+
+def nextpow(a: float, x: float) -> float:
+    """The smallest `a^n` not less than `x`, where `n` is a non-negative integer.
+
+    `a` must be greater than 1, and `x` must be greater than 0.
+    # Examples
+    ```jldoctest
+    julia> nextpow(2, 7)
+    8
+    julia> nextpow(2, 9)
+    16
+    julia> nextpow(5, 20)
+    25
+    julia> nextpow(4, 16)
+    16
+    ```
+    """
+    assert x > 0 and a > 1
+    if x <= 1:
+        return 1.0
+    n = math.ceil(math.log(x, a))
```

Comment:
I guess this is where the question from above belongs: why `math` and not `numpy`?

I don't really know Julia that well, or how to do docstrings in that context, but it'd be nice if here, like everywhere else, we could follow the `description, Parameters, Returns, Example` syntax :)

Reply:
`math` here gives the generic log in base `a`. I changed it into... `np.math.log` to have the same functionality :)